authorizations/authorization-ranger/src/main/java/org/apache/gravitino/authorization/ranger/RangerAuthorizationPlugin.java

```diff
@@ -0,0 +1,1006 @@
+    try {
+      rangerClient.getRole(roleName, RANGER_ADMIN_NAME, rangerServiceName);
+    } catch (RangerServiceException e) {
+      throw new AuthorizationPluginException(e);
+    }
+    return true;
+  }
+
+  private GrantRevokeRoleRequest createGrantRevokeRoleRequest(
+      String roleName, String userName, String groupName) {
+    Set<String> users;
+    if (userName == null || userName.isEmpty()) {
+      users = new HashSet<>();
+    } else {
+      users = new HashSet<>(Arrays.asList(userName));
+    }
+    Set<String> groups;
+    if (groupName == null || groupName.isEmpty()) {
+      groups = new HashSet<>();
+    } else {
+      groups = new HashSet<>(Arrays.asList(groupName));
+    }
+
+    GrantRevokeRoleRequest roleRequest = new GrantRevokeRoleRequest();
+    roleRequest.setUsers(users);
+    roleRequest.setGroups(groups);
+    roleRequest.setGrantor(RANGER_ADMIN_NAME);
+    roleRequest.setTargetRoles(new HashSet<>(Arrays.asList(roleName)));
+    return roleRequest;
+  }
+
+  /** Create a Ranger role if the role does not exist. */
+  private RangerRole createRangerRoleIfNotExists(String roleName) {
+    RangerRole rangerRole = null;
+    try {
+      rangerRole = rangerClient.getRole(roleName, RANGER_ADMIN_NAME, rangerServiceName);
+    } catch (RangerServiceException e) {
+      // ignore exception, If the role does not exist, then create it.
+    }
+    try {
+      if (rangerRole == null) {
+        rangerRole = new RangerRole(roleName, MANAGED_BY_GRAVITINO, null, null, null);
+        rangerClient.createRole(rangerServiceName, rangerRole);
+      }
+    } catch (RangerServiceException e) {
+      throw new RuntimeException(e);
+    }
+    return rangerRole;
+  }
+
+  /**
+   * Find the managed policy for the metadata object.
+   *
+   * @param metadataObject The metadata object to find the managed policy.
+   * @return The managed policy for the metadata object.
+   */
+  @VisibleForTesting
+  public RangerPolicy findManagedPolicy(MetadataObject metadataObject)
+      throws AuthorizationPluginException {
+    List<String> nsMetadataObj =
+        Lists.newArrayList(SecurableObjects.DOT_SPLITTER.splitToList(metadataObject.fullName()));
+    nsMetadataObj.remove(0); // skip `catalog`
+    Map<String, String> policyFilter = new HashMap<>();
+    Map<String, String> preciseFilterKeysFilter = new HashMap<>();
+    policyFilter.put(RangerDefines.SEARCH_FILTER_SERVICE_NAME, this.rangerServiceName);
+    policyFilter.put(SearchFilter.POLICY_LABELS_PARTIAL, MANAGED_BY_GRAVITINO);
+    for (int i = 0; i < nsMetadataObj.size(); i++) {
+      policyFilter.put(policyFilterKeys.get(i), nsMetadataObj.get(i));
+      preciseFilterKeysFilter.put(policyPreciseFilterKeys.get(i), nsMetadataObj.get(i));
+    }
+
+    try {
+      List<RangerPolicy> policies = rangerClient.findPolicies(policyFilter);
+
+      if (!policies.isEmpty()) {
+        // Because Ranger doesn't support the precise filter, Ranger will return the policy meets
```

Comment:
// Because Ranger only supports search the policies by key words, the results will contain many unnecessary data.
// For example, you search table1, Ranger return table1, table12, table13. RangerAuthorization should filter inaccurate data by hand.

Reply:
I rewrited these comments.